icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java

```diff
@@ -0,0 +1,311 @@
+// © 2020 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+
+package com.ibm.icu.impl.units;
+
+import com.ibm.icu.util.MeasureUnit;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.regex.Pattern;
+
+import static java.math.MathContext.DECIMAL128;
+
+public class UnitConverter {
+    private BigDecimal conversionRate;
+    private BigDecimal offset;
+
+    /**
+     * Constructor of `UnitConverter`.
+     * NOTE:
+     * - source and target must be under the same category
+     * - e.g. meter to mile --> both of them are length units.
+     *
+     * @param source          represents the source unit.
+     * @param target          represents the target unit.
+     * @param conversionRates contains all the needed conversion rates.
+     */
+    public UnitConverter(MeasureUnitImpl source, MeasureUnitImpl target, ConversionRates conversionRates) {
+        Convertibility convertibility = extractConvertibility(source, target, conversionRates);
+        assert (convertibility == Convertibility.CONVERTIBLE || convertibility == Convertibility.RECIPROCAL);
```

Comment:
This assert is failing for me I do get this assert failing for a global `ant check`.

I don't get an assert failure for this:
$ ant checkTest -Dtestclass='com.ibm.icu.dev.test.impl.UnitConverterTest'

I hope asserts are enabled for both? But then why is my "check" broken but my "checkTest" not?

I don't see this assert in C++ either, there's quite a bit of difference between the constructors.

Reply:
Turns out I had an old copy of unitsTest.txt in ./main/tests/core/out/bin/com/ibm/icu/dev/data/units/unitsTest.txt,
ant did not pick up the newer ./main/tests/core/src/com/ibm/icu/dev/data/cldr/units/unitsTest.txt

So that resolves my "ant check" problem. :-)